tools/releaser/main.go

```diff
@@ -0,0 +1,289 @@
+/* Copyright 2022 The Bazel Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// releaser is a tool for managing part of the process to release a new version of gazelle.
+//
+// This CLI is intended to be run directly with the go toolchain. It is not suitable for
+// running through 'bazel run' because it expectes to be able to mutate local files.
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"os/signal"
+	"regexp"
+	"strings"
+)
+
+const (
+	_tmpBzl = "tmp.bzl"
+)
+
+func main() {
+	ctx, cancel := signal.NotifyContext(context.Background(), os.Interrupt)
+	defer cancel()
+	if err := cli(ctx, os.Stderr, os.Args[1:]); err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
+	}
+}
+
+func cli(ctx context.Context, stderr *os.File, args []string) error {
+	if len(args) > 1 || len(args) == 1 && strings.HasPrefix(strings.TrimLeft(args[0], "-"), "h") {
+		fmt.Println(`usage: go run tools/releaser/main.go
+
+This utility is intended to handle many of the steps to release a new version.
+
+Only run it from the root of the bazel-gazelle repository.`)
+		return nil
+	}
+
+	var verbose bool
+	if len(args) == 1 && strings.HasPrefix(strings.TrimLeft(args[0], "-"), "v") {
+		verbose = true
+	}
+
+	if verbose {
+		fmt.Println("Running initial go update commands")
+	}
+	initialCommands := []struct {
+		cmd  string
+		args []string
+	}{
+		{cmd: "go", args: []string{"get", "-t", "-u", "./..."}},
+		{cmd: "go", args: []string{"mod", "tidy", "-compat=1.16"}},
+		{cmd: "go", args: []string{"mod", "vendor"}},
+		{cmd: "find", args: []string{"vendor", "-name", "BUILD.bazel", "-delete"}},
+	}
+	for _, c := range initialCommands {
+		if out, err := exec.CommandContext(ctx, c.cmd, c.args...).CombinedOutput(); err != nil {
+			fmt.Println(string(out))
+			return err
+		}
+	}
+
+	workspace, err := os.ReadFile("WORKSPACE")
+	if err != nil {
+		return err
+	}
+
+	workspaceBuffer := bytes.NewBuffer(workspace)
+	workspaceWithoutDirectives := new(bytes.Buffer)
+
+	if verbose {
+		fmt.Println("Preparing temporary WORKSPACE without gazelle directives.")
+	}
+	directiveStart, err := getWorkspaceWithouthDirectives(workspaceBuffer, workspaceWithoutDirectives)
```

Comment:
Can we always write the `gazelle:repository` directive at the bottom of `WORKSPACE` file? So we don't need to remember `directiveStart`